### ISSUE TITLE
Ensemble Model only marginally faster than separate single-model calls -- is this expected?

### ISSUE THREAD

**Issue by corentin87** [triton-inference-server/server#8740]
Hi everyone,

I am evaluating whether a Triton **ensemble pipeline** (with BLS for conditional
logic) can outperform a **two-call single-model** approach for an RT-DETR
detection + ReID re-identification pipeline. Server-side preprocessing (Python/DALI)  and postprocessing (Python) are
significantly slower** than keeping pre/postprocessing on the client. I'd like to
understand if this is expected behavior or if there are configuration
improvements I'm missing.

## Setup

- **Triton Server:** custom image with Python, TRT and Dali backend  from 2.66.0 (`nvcr.io/nvidia/tritonserver:26.02-py3`) 
- **GPU:** NVIDIA RTX A2000
- **Models:**

| Model                        | Backend        | Precision | Max Batch |
|------------------------------|----------------|-----------|-----------|
| `rt-detr`                    | TensorRT (GPU) | FP16      | 8         |
| `reid`                       | TensorRT (GPU) | FP16      | 256       |
| `detect_preprocessing`       | Python (CPU) or DALI (GPU) | — | — |
| `detect_postprocessing_reid` | Python (CPU), BLS | —      | —         |

Pre/Post processing steps in details:

- **detect_preprocessing** is doing:
    - resizing
   -  RGB conversion
  -  normalisation
- **detect_postprocessing_reid** is doing:
  -  post processing on the bounding boxes
  -  cropping the detections from the original image
  -    calling ReID (sync)
  -  normalising the feature vectors

## Pipeline Architectures

**Single mode:** The client makes two separate async gRPC calls per frame and
handles all pre/post-processing locally:

Client (pre/post-processing) → gRPC 1 → rt-detr → Client (post/pre) → gRPC 2 → reid → Client (post)

**Ensemble mode:** A single async gRPC call drives the full pipeline server-side. BLS
is used inside `detect_postprocessing_reid` so that `reid` is only invoked when
detections are present (conditional logic that the static ensemble DAG cannot
express):
raw_image → detect_preprocessing → rt-detr → detect_postprocessing_reid (BLS) → reid → outputs

Ensemble config.pbtxt:

```name: "ensemble_model"
platform: "ensemble"
max_batch_size: 8

input [
  {
    name: "raw_image"
    data_type: TYPE_UINT8
    dims: [ -1, -1, 3 ]
  },
  {
    name: "confidence_threshold"
    data_type: TYPE_FP32
    dims: [ 1]
  }
]

output [
  {
    name: "detection_boxes_out"
    data_type: TYPE_FP32
    dims: [ 4 ]
  },
  {
    name: "detection_confidences_out"
    data_type: TYPE_FP32
    dims: [ 1 ]
  },
  {
    name: "detection_labels_out"
    data_type: TYPE_FP32
    dims: [ 1 ]
  },
  {
    name: "descriptors_out"
    data_type: TYPE_FP32
    dims: [ 128 ]
  }
]

ensemble_scheduling {
  step [
    {
      model_name: "detect_preprocessing"
      model_version: -1
      input_map {
        key: "image"
        value: "raw_image"
      }
      output_map {
        key: "image"
        value: "image"
      }
    },
    {
      model_name: "rt-detr"
      model_version: -1
      input_map {
        key: "inputs"
        value: "image"
      }
      output_map {
        key: "pred_boxes"
        value: "detection_boxes"
      }
      output_map {
        key: "pred_logits"
        value: "detection_logits"
      }
    },
    {
      model_name: "detect_postprocessing_reid"
      model_version: -1
      input_map {
        key: "detection_logits"
        value: "detection_logits"
      }
      input_map {
        key: "detection_boxes"
        value: "detection_boxes"
      }
      input_map {
        key: "image"
        value: "raw_image"
      }
      input_map {
        key: "confidence_threshold"
        value: "confidence_threshold"
      }
      output_map {
        key: "detection_boxes"
        value: "detection_boxes_out"
      },
      output_map {
        key: "detection_confidences"
        value: "detection_confidences_out"
      },
      output_map {
        key: "detection_labels"
        value: "detection_labels_out"
      }
      output_map {
        key: "descriptors"
        value: "descriptors_out"
      }
    }
  ]
}
```

Based on initial experiments, `detect_preprocessing` running as a Python backend
model appeared to be a bottleneck. To evaluate this, preprocessing was tested in
three variants:
- **Client-side** — preprocessing on the client before sending to the ensemble
  (not part of the ensemble graph)
- **Python backend** — preprocessing on the server using OpenCV (CPU)
- **DALI backend** — preprocessing on the server using NVIDIA DALI (GPU)

### Pipeline Throughput (FPS) — concurrency sweep, no dynamic batching
| Pipeline             | Pre-processing  | c=1       | c=4       | c=8       |
|----------------------|-----------------|-----------|-----------|-----------|
| Single               | Client          | 8.0 FPS   | 74.3 FPS  | 73.0 FPS  |
| Ensemble             | Client          | 11.6 FPS  | 77.4 FPS  | 75.3 FPS  |
| Ensemble             | Python Backend  | 13.0 FPS  | 38.0 FPS  | 61.4 FPS  |
| Ensemble             | DALI Backend    | 12.7 FPS  | 39.7 FPS  | 67.7 FPS  |

### Pipeline Throughput with dynamic batching enabled (concurrency=8)
| Configuration        | No DM     | DM        | Delta |
|----------------------|-----------|-----------|-------|
| Single + Client      | 73.0 FPS  | 73.6 FPS  | +1%   |
| Ensemble + Client    | 75.3 FPS  | 78.9 FPS  | +5%   |
| Ensemble + Python    | 61.4 FPS  | 56.2 FPS  | −8%   |
| Ensemble + DALI      | 67.7 FPS  | 66.2 FPS  | −2%   |

### GPU utilisation (concurrency=8)
| Configuration        | GPU Util (%) | CPU (%) |
|----------------------|--------------|---------|
| Single + Client      | 51–54        | 41–43   |
| Ensemble + Client    | 54–55        | 52      |
| Ensemble + Python    | 44–46        | 57      |
| Ensemble + DALI      | 52–54        | 60–62   |

Below are the plots of the different average times. (dynamic batching enabled) 

<img width="2400" height="2025" alt="Image" src="https://github.com/user-attachments/assets/752cb4dc-5695-4ce7-8a97-9cfce1c3f437" />

Instance Count sweep for Python and Dali backend

<img width="1575" height="1950" alt="Image" src="https://github.com/user-attachments/assets/2b8924e1-8702-408e-935b-ebccbca16f79" />

<img width="1575" height="1950" alt="Image" src="https://github.com/user-attachments/assets/bc466a15-b058-46de-b2d2-69598409ca36" />

### Model Analyzer

To find the optimal configuration for each ensemble variant, I attempted to use
Model Analyzer. However, I encountered two blockers:

1. **Ensemble + BLS profiling is not supported.** According to the documentation,
   Model Analyzer cannot profile an ensemble model that includes a BLS call. I opened [an issue](https://github.com/triton-inference-server/model_analyzer/issues/1015) requesting guidance but have not received a response yet.

2. **Triton Server crashes during model reload.** After removing the BLS call to
   profile the remaining models, Model Analyzer repeatedly crashes the remote
   Triton Server when it attempts to reload models with updated configurations. Profiling works fine with
   `reload_model_disable: true` but i can't automatically sweep. I reported this in [a separate issue](https://github.com/triton-inference-server/model_analyzer/issues/1018) also without response so far.

## Observations

1. **Server-side preprocessing is significantly slower.** The average request
   time is ~10x higher with the Python backend and ~6x higher with the DALI
   backend compared to client-side preprocessing. Queue times are also
   substantially elevated.

2. **Increasing instance count helps but does not close the gap.** Scaling the
   Python or DALI backend from 1 to 2 or 4 instances reduces the request time
   by ~1.5–2x, but it remains well above the client-side preprocessing baseline.

3. **Ensemble + Client is only marginally faster than Single** at high
   concurrency (75–79 FPS vs 73–74 FPS, a 3–7% improvement).

4** I was also hopping to see a benefit in having the post processing image cropping of all the bounding boxes in the Ensemble rather than having to pass all the cropped images (~20/per frame in the current video test) from the client to the model.

Is it expected that an ensemble with BLS shows only marginal improvement
over separate model calls at high concurrency? Our hypothesis was that
eliminating the second gRPC round-trip and keeping data server-side would
yield a larger speedup, but the benefit seems to minimal.

Any guidance on whether these results are in line with what's expected for this
type of pipeline, or if there are optimizations I am missing, would be very
appreciated.

**Comment by gjegoj** [comment]
@corentin87 
Hi!
I got exactly the same result when testing ensembles for our task. In our case, ensembles always performed 10-15% worse than individual queries. We've now come up with a solution that NVIDIA doesn't recommend: sending multiple queries from the client and then performing some post-processing on the results.

**Comment by corentin87** [comment]
Glad we are not the only one :) . Where did you see you can't send multiple queries? Are you not just batching the queries and sending them to the server?
